### PR TITLE
Implement new Interning Checker semantics 

### PR DESCRIPTION
**Pull request by smillst**
(A PR in to the defaultyChanges branch)

New semantics are explained: https://docs.google.com/document/d/1w7rM0fIqP4CMPHddce7sG-SmFAou0Juj0Hqeml6Nkbk/edit

That Google doc will be integrated into the manual in another pull request.

### CODE REVIEW

**Comment by smillst** [comment]
The demos are failing because they still need to be adapted to the new semantics.  @mernst is working on this.

**Comment by smillst** [comment]
@wmdietl The test are passing now.

File: checker/src/main/java/org/checkerframework/checker/interning/InterningAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -65,6 +72,62 @@ public InterningAnnotatedTypeFactory(BaseTypeChecker checker) {
...
+            MethodSymbol methodElt = (MethodSymbol) type.getElement();
+
+            if (methodElt == null
+                    || methodElt.getKind() != ElementKind.CONSTRUCTOR
+                    || methodElt.owner.getKind() == ElementKind.ENUM) {
+                // Annotate method returns, not constructors.
+                scan(type.getReturnType(), p);
+            }
+            if (type.getReceiverType() != null
+                    && typeFactory.getDeclAnnotation(type.getElement(), InternMethod.class)
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/interning/InterningAnnotatedTypeFactory.java]: You could use `methodElt` again instead of calling `type.getElement()`.

Comment (wmdietl) [reply checker/src/main/java/org/checkerframework/checker/interning/InterningAnnotatedTypeFactory.java]: A comment would be useful to understand what should happen with methods that have the `InternMethod` annotation.

File: checker/src/main/java/org/checkerframework/checker/interning/InterningAnnotatedTypeFactory.java [hunk 2]
```diff
@@ -65,6 +72,62 @@ public InterningAnnotatedTypeFactory(BaseTypeChecker checker) {
...
+        }
+
+        @Override
+        public Void visitExecutable(AnnotatedExecutableType type, Void p) {
+            MethodSymbol methodElt = (MethodSymbol) type.getElement();
+
+            if (methodElt == null
+                    || methodElt.getKind() != ElementKind.CONSTRUCTOR
+                    || methodElt.owner.getKind() == ElementKind.ENUM) {
+                // Annotate method returns, not constructors.
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/interning/InterningAnnotatedTypeFactory.java]: `... not constructors or enums`? There is a check for enums, too.

File: checker/src/main/java/org/checkerframework/checker/interning/InterningVisitor.java [hunk 3]
```diff
@@ -289,32 +291,58 @@ protected void checkConstructorResult(
...
+     */
+    private boolean checkCreationOfInternedObject(
+            NewClassTree newInternedObject, AnnotatedExecutableType constructor) {
+        if (constructor.getReturnType().hasAnnotation(Interned.class)) {
+            return true;
+        }
+        if (getCurrentPath() != null
+                && getCurrentPath().getParentPath() != null
+                && getCurrentPath().getParentPath().getParentPath() != null) {
+            Tree parent = getCurrentPath().getParentPath().getParentPath().getLeaf();
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/interning/InterningVisitor.java]: Can you add a comment why there are two parents to climb?
I assume this is to handle `myIntern(new MyInternType())`. Why is it not just one level to climb to get from the `new` tree to the `myIntern` tree?

Comment (wmdietl) [reply checker/src/main/java/org/checkerframework/checker/interning/InterningVisitor.java]: Oh, or is it for `new MyInternType().intern()`? So a comment would definitely be useful.

File: framework/src/main/java/org/checkerframework/common/aliasing/AliasingAnnotatedTypeFactory.java [hunk 4]
```diff
@@ -55,7 +57,13 @@ public AliasingAnnotatedTypeFactory(BaseTypeChecker checker) {
     // type qualifier anymore.
     @Override
     protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
-        return getBundledTypeQualifiersWithoutPolyAll(MaybeLeaked.class);
+        return new LinkedHashSet<>(
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/common/aliasing/AliasingAnnotatedTypeFactory.java]: Is this change related to interning?

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 5]
```diff
@@ -1157,6 +1157,12 @@ public AnnotatedTypeMirror getAnnotatedTypeFromTypeTree(Tree tree) {
         return qualifierUpperBounds.getBoundQualifiers(type);
     }
 
+    /** Returns the set of qualifiers that are the upper bounds for a use of the type. */
+    public Set<AnnotationMirror> getTypeDeclarationBounds(TypeElement typeElement) {
+        return qualifierUpperBounds.getBoundQualifiers(
+                AnnotatedTypeMirror.createType(typeElement.asType(), this, true));
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: It seems rather costly to create an `AnnotatedTypeMirror` here. `QualifierUpperBounds.getBoundQualifiers` then deconstructs it into an Element again very soon
https://github.com/typetools/checker-framework/blob/defaultyChanges/framework/src/main/java/org/checkerframework/framework/type/QualifierUpperBounds.java#L122

Maybe an `Element` version can be added there?

File: framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 6]
```diff
@@ -3390,20 +3390,33 @@ private boolean checkMemberReferenceReceivers() {
...
-            overriddenReceiver.addAnnotations(overridden.getReceiverType().getAnnotations());
-            if (!atypeFactory
-                    .getTypeHierarchy()
-                    .isSubtype(overriddenReceiver, overrider.getReceiverType().getErased())) {
-                FoundRequired pair =
-                        FoundRequired.of(overrider.getReceiverType(), overridden.getReceiverType());
+            Set<AnnotationMirror> overriderAnnos = overriderReceiver.getAnnotations();
+            Set<AnnotationMirror> overriddenAnnos = overriddenReceiver.getAnnotations();
+            if (!qualifierHierarchy.isSubtype(overriddenAnnos, overriderAnnos)) {
+                // If the overriding receiver annotations are the same as the bound for uses of the
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java]: This comment would be better within the `return true` block below.
I don't quite understand the description though. Can you clarify why this makes sens?

File: checker/src/main/java/org/checkerframework/checker/interning/InterningAnnotatedTypeFactory.java (hunk 2 above)
Comment (smillst) [reply checker/src/main/java/org/checkerframework/checker/interning/InterningAnnotatedTypeFactory.java]: Enum check was wrong.  Thanks!

File: checker/src/main/java/org/checkerframework/checker/interning/InterningVisitor.java (hunk 3 above)
Comment (smillst) [reply checker/src/main/java/org/checkerframework/checker/interning/InterningVisitor.java]: I added a comment below.
